### ISSUE TITLE
nvim_open_win(): support regular windows

### ISSUE THREAD

**Issue by andmis** [neovim/neovim#18560]
### Feature already in Vim?

No

### Feature description

I would like to be able to do the following:
- When calling `:help some-thing`, if there is no current help window, I would like a full-height vertical split to be created on the right side of the editor, and the help window to be the right-hand split. If there is an existing help window, I would like it to be reused.
- When calling `:Man some-thing`, I would like identical behavior.

Both of those are possible today because both `:help` and `:Man` have built-in logic to reuse existing `:help`/`:Man` windows. But what if I want `:help some-thing` to reuse either the `ft=help` or the `ft=man` window if _either_ exists? Or what if I want to open `:help` in the current window? (Today there is the janky tip `help-curwin` for how to do this...)

---

I think the main problem is that there are two distinct things that commands like `:help`, `:copen`, `:Man`, as well as tons of plugin commands such as `:G log` in `vim-fugitive`, are actually doing:
1. Picking a window to put their contents into, potentially creating a new split.
2. Opening the appropriate content buffer in the chosen window.

Right now, if I want to have manual control over step (1), there is often no straightforward way to do that. And because the Ex commands are often the main/only interface to the operations being performed ("open the help page for "), it's not even clear _how_ to give users control over (1). You can create global options/settings, but that sucks for (reasons).

If I am determined to control where splits go, as of today I have to resort to ad-hoc hacks which vary by plugin / window type (for example, read `:help help-curwin`).

---

I think a better interface to move towards for operations like "open the help page for some-thing" or "show the Git status" would be:
- Provide a Lua function (per operation) that opens a buffer implementing the operation in either (1) a given `win_id` or (2) a new split described somehow via a function argument (to avoid creating a split with a scratch/unused buffer).
  - The signature would look something like `help_open(help_str, win_id_or_split_descr, opts...)`.
- Provide a convenience Lua function (per operation) that computes the window or split-type into which the operation wants to go, by default.
  - For example, `help_default_window` might return "the current help `win_id` if there is one, else some representation of `the upper half of a horizontal split`".
- Perhaps provide convenience functions composing these two.

**Comment by andmis** [comment]
Just to flesh this out a bit more, we would probably need to define a `win_open_spec` record type, which allows one to express where a buffer should be opened. This can encode:
- just a `win_id`, saying "open the buffer in this `win_id`, or
- a `win_id`, a split type (horizontal or vertical), and a direction (`:leftabove` or `:rightbelow`), and
- perhaps an enum making it explicit which of the above two choices is desired.

Then we would provide a functions in Lua which, given a `buf` and a `win_open_spec`, opens the `buf` in that `win_open_spec`, either using the given window or creating a new split.

**Comment by justinmk** [comment]
> I think a better interface to move towards for operations like "open the help page for some-thing" or "show the Git status" would be:
> 
> * Provide a Lua function (per operation) ...

- Let's ignore the implementation ("Lua function...") for now. What is the actual interface you expect, in terms of Ex commands? How would `:help` look differently? `:vert help` opens a vertical window. Are you looking for a way to fine-tune the window layout? That's a big topic but let's start with what the interface would look like.
    - https://github.com/neovim/neovim/issues/3933
- The issue title mentions "special buffers", why is this not applicable to all kinds of buffers/windows?

**Comment by andmis** [comment]
I would like to be able to write something like the following:

```
function my_help_opener(help_str)
  local win_id
  -- Iterate over the existing windows. If any of them has `ft=help` or `ft=man`, then
  -- set `win_id` to point to that window. This is straightforward today.

  local win_open_spec
  if win_id then
    win_open_spec = {
      win_id = win_id,
      type = 'open_in_existing',
    }
  else
    win_open_spec = {
      type = 'vsplit_in_editor',
      dir = 'right'
    }
  end

  nvim.api.open_buffer(nvim.api.make_help_buffer(help_str), win_open_spec)
end
```

In terms of the Ex interface, I would like it to be trivial to create my own Ex command `Help`, so that `:Help my_help_str` just calls `my_help_opener("my_help_str")`. I have never tried to define an Ex command from Lua so I don't know how easy/hard that is today.

**Comment by andmis** [comment]
This does apply to all kinds of buffers/windows, but from the user's perspective, we already have good control over windows for "normal" buffers (the `<C-w>` suite of commands, and `:split` and `:vsplit`, etc.).

The reason I am focusing on "special" buffers is that the end-user issue I am pointing at is that Ex commands, both built-in (`:help`, `:copen`, etc.) and defined in plugins (`:Git`), make their own decisions about how/where to open/split windows for their content, and the user has no control over this.

---

Just to make it super-concrete, if I have a large monitor, and use Neovim in a full-screen window, it makes no sense at all to have `:help` open in a full-width horizontal split which is like, 300 columns wide. Same for things like `:Git`, which also opens a horizontal full-width split.

**Comment by andmis** [comment]
The current default behavior of `:help`:

<img width="1840" alt="image" src="https://user-images.githubusercontent.com/5308024/168323962-007d15fb-0488-491f-a2d3-21e3c1fc738e.png">

**Comment by justinmk** [comment]
> nvim.api.open_buffer(nvim.api.make_help_buffer(help_str), win_open_spec)

Ah, you want to control window layout with `nvim_create_buf()` or `nvim_open_win()`

What's missing from `nvim_open_win()` ?

**Comment by andmis** [comment]
> What's missing from `nvim_open_win()` ?

Can you open non-floats with `nvim_open_win()`? I thought this only works for floats (`'relative'`) and external windows.

**Comment by andmis** [comment]
The other use case is, suppose that as a plugin author, I am going to expose some functionality in a special buffer (like `:Git`) and I want to give users or other plugin authors programmatic control over where that buffer will appear. There is no noun today for "where to put that buffer", that covers both (1) putting it into an existing window, and (2) spawning a new split for it.

As a result, plugin authors either don't expose that optionality to users (since there is no idiom for doing so), or if they do, it's in the form of global config options (but this is pretty rare).

**Comment by justinmk** [comment]
so the request here is for `nvim_open_win()` to support normal windows.

**Comment by justinmk** [comment]
duplicate https://github.com/neovim/neovim/issues/14315

**Comment by andmis** [comment]
Not quite, the request is a function that takes a buffer and a "location" to open the buffer, where the "location" can either be a config for a new window or an existing window ID.

The reason for the interface to look this way is to encourage plugin authors to factor their logic into functional computations of (1) what to open and (2) where to open it. Otherwise, as a plugin author, if I want to give a user control over where my content gets opened, it can be tempting to expose two functions `open_plugin_stuff_in_existing_window` and `open_plugin_stuff_in_new_split`, which is a bad outcome because people will define the shape of the arguments to `open_plugin_stuff_in_new_split` in inconsistent ways, and may fail to expose `make_buffer_with_plugin_stuff_in_it` that I can pass to `nvim_open_win` myself, which is what I actually want.

Basically the ask is to set up the API to nudge plugin authors into this way of factoring the logic for displaying TUI buffers.

**Comment by msaher** [comment]
@andmis For what its worth this sounds very similar to `display-buffer-alist` in emacs. It's an abstraction that allows users to override defaults and determine how certain buffers should be displayed. For instance, a user may change the default behavior of `M-x shell` to make it use the current window instead of opening up in a new split. In neovim we would be able to configure how window-opening commands like `:term`' and `:Man`show up.

**Comment by msaher** [comment]
> @andmis For what its worth this sounds very similar to `display-buffer-alist` in emacs. It's an abstraction that allows users to override defaults and determine how certain buffers should be displayed. For instance, a user may change the default behavior of `M-x shell` to make it use the current window instead of opening up in a new split. In neovim we would be able to configure how window-opening commands like `:term` and `:Man`show up.

@justinmk thoughts on this? Of course #14315 has to be closed first, but its something to consider afterwards.

**Comment by justinmk** [comment]
UI extensibility is an ongoing topic, in particular window layout is already controllable by clients implemented `:help ui-mutligrid`. Emacs has nothing like this.
